### PR TITLE
Feature/bar chart

### PR DESCRIPTION
**Pull request by totzk9**
# 📲 What #

* Implemented BarChart of type Vertical Stacked and Normal.

# 🤔 Why #

* New feature.

# 🛠 How #
 
* In BarChart
  - In drawData(...), draw a different path whose draw style is fill if type is BarChartType.VERTICAL or BarChartType.VERTICAL_STACKED.
  - Added composable preview for BarChartType.VERTICAL.

# ✅ Acceptance criteria #

- [x] Composable preview added
- [x] Added UI Tests

# ⏰ TODO #

* None

# 🔗 References #
Screen Shot 2021-04-26 at 3 02 18 PM
Screen Shot 2021-04-26 at 3 02 30 PM
image

### CODE REVIEW

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt [hunk 1]
```diff
@@ -1,18 +1,277 @@
...
+ */
+private fun <T> DrawScope.drawData(barSeries: BarSeries<T>) {
+    barSeries.data.forEachIndexed { i, chartData ->
+        val point = getDataPoint(i, chartData, size.height)
+        drawRect(
+                topLeft = Offset(point.x - 24, point.y),
+                size = Size(width = 48f, height = size.height - point.y),
+                color = barSeries.color,
+                style = Stroke(barSeries.width.toFloat()),
+//              brush = Brush.linearGradient(colors = listOf(barSeries.areaColor!!))
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: remove if not used

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt [hunk 2]
```diff
@@ -1,18 +1,277 @@
...
+                tmpLegend
+            }
+    ) {
+        Canvas(
+                modifier = modifier
+                        .weight(5f)
+                        .padding(
+                                start = (mMaxVal.toString().length * 8).dp,
+                                top = 10.dp,
+                                end = 10.dp,
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: create a diff function for this

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt [hunk 3]
```diff
@@ -1,18 +1,277 @@
...
+/**
+ * Draw line chart grids.
+ *
+ * @param count Grid count.
+ * @param color Grid color.
+ */
+private fun DrawScope.drawGrids(count: Int, color: Color) {
+    val startPoint = origin()
+//    val pathEffect = PathEffect.dashPathEffect(intervals = floatArrayOf(10f, 10f))
+
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: remove if not used

Comment (rexendy) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: remove all commented codes if not used

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt [hunk 4]
```diff
@@ -1,18 +1,277 @@
...
+            )
+            val keyOffset = Offset(
+                    (i + 1) * mHGap,
+                    startPoint.y + AXIS_VALUES_FONT_SIZE + 10f
+            )
 
+            it.nativeCanvas.drawText(
+                    (valuePerGrid * (i + 1)).toInt().toString(),
+                    valOffset.x,
+                    valOffset.y,
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: if possible make a function for this

File: kwarchart/src/main/java/com/kwarchart/android/model/BarSeries.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package com.kwarchart.android.model
+
+import androidx.compose.ui.graphics.Color
+import com.kwarchart.android.enum.BarChartType
+
+/**
+ * LineChart data.
+ */
+data class BarSeries<T>(
+        /**
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/model/BarSeries.kt]: apply proper spacing
/** pls align it to its closing tag

File: sample/src/main/res/layout/fragment_bar.xml [hunk 6]
```diff
@@ -6,17 +6,22 @@
     android:layout_height="match_parent"
     tools:context=".ui.bar.BarFragment">
 
-    <TextView
-        android:id="@+id/text_dashboard"
+    <androidx.compose.ui.platform.ComposeView
+        android:id="@+id/chartBar"
```
Comment (rexendy) [inline sample/src/main/res/layout/fragment_bar.xml]: pls use snake case convention for xml id naming

File: sample/src/main/java/com/kwarchart/sample/ui/bar/BarFragment.kt [hunk 7]
```diff
@@ -4,21 +4,26 @@ import android.os.Bundle
...
-import androidx.lifecycle.Observer
 import androidx.lifecycle.ViewModelProvider
 import com.kwarchart.android.chart.BarChart
 import com.kwarchart.android.enum.LegendPosition
 import com.kwarchart.sample.databinding.FragmentBarBinding
 
-class BarFragment : Fragment() {
+
+
+
```
Comment (rexendy) [inline sample/src/main/java/com/kwarchart/sample/ui/bar/BarFragment.kt]: remove extra line breaks

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt [hunk 8]
```diff
@@ -166,15 +197,20 @@ private fun DrawScope.drawGrids(count: Int, color: Color) {
...
-                start = Offset(i * mHGap, 0f),
-                end = Offset(i * mHGap, size.height),
-//            pathEffect = pathEffect
-        )
+//        drawLine(
+//                color = color,
+//                start = Offset(i * mHGap, 0f),
+//                end = Offset(i * mHGap, size.height),
+////            pathEffect = pathEffect
+//        )
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: remove if not used

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt [hunk 9]
```diff
@@ -288,15 +288,36 @@ fun BarChartPreview() {
...
+                            legend = "Spent"
                     )
-            )
+            ),
+            type = VERTICAL,
+//            type = VERTICAL_STACKED,
+
+            ///TODO
+//            type = HORIZONTAL,
+//            type = HORIZONTAL_STACKED,
```
Comment (rexendy) [inline kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: remove if not used

**Review by rexendy — CHANGES_REQUESTED** [review]
Hello Ty

- pls check spacing indentation and line breaks
coordindate it with Rick to have same format
- pls add all static text to strings.xml
- pls remove all commented codes that are not used
- if possible, make your functions not too long ( optimize only for those
functions needed)

Nice test. thank you

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt (hunk 9 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt (hunk 8 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: sample/src/main/java/com/kwarchart/sample/ui/bar/BarFragment.kt (hunk 7 above)
Comment (rickimaru18) [reply sample/src/main/java/com/kwarchart/sample/ui/bar/BarFragment.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: sample/src/main/res/layout/fragment_bar.xml (hunk 6 above)
Comment (rickimaru18) [reply sample/src/main/res/layout/fragment_bar.xml]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: kwarchart/src/main/java/com/kwarchart/android/model/BarSeries.kt (hunk 5 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/model/BarSeries.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt (hunk 4 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt (hunk 3 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt (hunk 2 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

File: kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt (hunk 1 above)
Comment (rickimaru18) [reply kwarchart/src/main/java/com/kwarchart/android/chart/BarChart.kt]: Let me handle this one... Will fix it in HORIZONTAL BarChart PR.

**Comment by rickimaru18** [comment]
@totzk9 @rexendy
Bugs found:
1. Stack display incorrect. Should stack to Y-axis, not Z-axis.
2. Incorrect display when 3 or more BarSeries .

I'll just fix this when I implement HORIZONTAL BarChart.

**Comment by rickimaru18** [comment]
To be continued in horizontal_bar_chart branch.

**Review by rickimaru18 — APPROVED** [review]

**Comment by rickimaru18** [comment]
> Hello Ty
> 
>     * pls check spacing indentation and line breaks
>       coordindate it with Rick to have same format
> 
>     * pls add all static text to strings.xml
> 
>     * pls remove all commented codes that are not used
> 
>     * if possible, make your functions not too long ( optimize only for those
>       functions needed)
> 
> 
> Nice test. thank you

@rexendy 
CC: @totzk9 

Please see 1717eb9 and 37d0304 :bow:

**Comment by rexendy** [comment]
> @totzk9 @rexendy
> Bugs found:
> 
> 1. Stack display incorrect. Should stack to Y-axis, not Z-axis.
> 2. Incorrect display when 3 or more BarSeries .
> 
> I'll just fix this when I implement HORIZONTAL BarChart.

nice catch @rickimaru18

**Review by rexendy — APPROVED** [review]

**Comment by rexendy** [comment]
@rickimaru18 pls merge it to develop branch.

Thank you

**Pull request merged**